pydantic_ai_slim/pydantic_ai/ui/event_stream.py

```diff
@@ -0,0 +1,534 @@
+
+    async def handle_part_start(self, event: PartStartEvent) -> AsyncIterator[EventT]:
+        """Handle a PartStartEvent.
+
+        Args:
+            event: The PartStartEvent.
+        """
+        part = event.part
+        previous_part_kind = event.previous_part_kind
+        match part:
+            case TextPart():
+                async for e in self.handle_text_start(part, follows_text=previous_part_kind == 'text'):
+                    yield e
+            case ThinkingPart():
+                async for e in self.handle_thinking_start(part, follows_thinking=previous_part_kind == 'thinking'):
+                    yield e
+            case ToolCallPart():
+                async for e in self.handle_tool_call_start(part):
+                    yield e
+            case BuiltinToolCallPart():
+                async for e in self.handle_builtin_tool_call_start(part):
+                    yield e
+            case BuiltinToolReturnPart():
+                async for e in self.handle_builtin_tool_return(part):
+                    yield e
+            case FilePart():
+                async for e in self.handle_file(part):
+                    yield e
+
+    async def handle_part_delta(self, event: PartDeltaEvent) -> AsyncIterator[EventT]:
+        """Handle a PartDeltaEvent.
+
+        Args:
+            event: The PartDeltaEvent.
+        """
+        delta = event.delta
+        match delta:
+            case TextPartDelta():
+                async for e in self.handle_text_delta(delta):
+                    yield e
+            case ThinkingPartDelta():
+                async for e in self.handle_thinking_delta(delta):
+                    yield e
+            case ToolCallPartDelta():
+                async for e in self.handle_tool_call_delta(delta):
+                    yield e
+
+    async def handle_part_end(self, event: PartEndEvent) -> AsyncIterator[EventT]:
+        """Handle a PartEndEvent.
+
+        Args:
+            event: The PartEndEvent.
+        """
+        part = event.part
+        next_part_kind = event.next_part_kind
+        match part:
+            case TextPart():
+                async for e in self.handle_text_end(part, followed_by_text=next_part_kind == 'text'):
+                    yield e
+            case ThinkingPart():
+                async for e in self.handle_thinking_end(part, followed_by_thinking=next_part_kind == 'thinking'):
+                    yield e
+            case ToolCallPart():
+                async for e in self.handle_tool_call_end(part):
+                    yield e
+            case BuiltinToolCallPart():
+                async for e in self.handle_builtin_tool_call_end(part):
+                    yield e
+            case BuiltinToolReturnPart() | FilePart():
+                # These don't have deltas, so they don't need to be ended.
+                pass
+
+    async def handle_text_start(self, part: TextPart, follows_text: bool = False) -> AsyncIterator[EventT]:
```

Comment:
I should make it more clear which of these methods require you to call super when overriding them, and which are "leaves" and don't need it.